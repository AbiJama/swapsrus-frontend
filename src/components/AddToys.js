import React, { useState } from "react";
import axios from "axios";
import Alert from "./Alert";
import "../styles/add-toys.css";

function AddToys(pristine) {
  const initialState = {
    title: "",
    type: "books",
    condition: "new",
    ageRange: "",
    description: "",
    borrowPeriod: "select borrow period",
    postcode: "",
    image: "",
  };

  const [fields, setFields] = useState(initialState);
  const [alert, setAlert] = useState({ message: "", isSuccess: false });

  const handleAddToys = async (event) => {
    event.preventDefault();
    setAlert({ message: "", isSuccess: false });

    if (
      !fields.title ||
      !fields.ageRange ||
      !fields.description ||
      !fields.borrowPeriod
    ) {
      setAlert({
        message: "Please fill out all required fields.",
        isSuccess: false,
      });
      return;
    }

    try {
      const response = await axios.post("http://localhost:4000/toys", fields);
      setAlert({
        message: "The item has been added successfully!",
        isSuccess: true,
      });
      setFields(initialState);
    } catch (error) {
      setAlert({
        message: "Server error, please try again later.",
        isSuccess: false,
      });
    }
  };
  const handleFieldChange = (event) => {
    const { name, value } = event.target;
    setFields((prevFields) => ({
      ...prevFields,
      [name]: value,
    }));
  };

  return (
    <div className="add-toys">
      <h2>Add Toys</h2>
      <form className="add-toy-form" onSubmit={handleAddToys}>
        <label htmlFor="title">
          Toy Title:
          <br />
          <br />
          <input
            id="title"
            name="title"
            placeholder="Enter title here..."
            value={fields.title}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="type">
          Type:
          <br />
          <br />
          <select
            id="type"
            name="type"
            value={fields.type}
            onChange={handleFieldChange}
          >
            <option value="Choose Category">Choose Category</option>
            <option value="Books">Books</option>
            <option value="Pre-school">Pre-school</option>
            <option value="Indoor">Indoor</option>
            <option value="Outdoor">Outdoor</option>
          </select>
        </label>
        <label htmlFor="condition">
          Condition:
          <br />
          <br />
          <select
            id="condition"
            name="condition"
            value={fields.condition}
            onChange={handleFieldChange}
          >
            <option value="Select Condition">Select Condition</option>
            <option value="Brand New">Brand New</option>
            <option value="Like New">Like New</option>
            <option value="Good">Good</option>
            <option value="Fair">Fair</option>
            <option value="Defected">Defected</option>
          </select>
        </label>
        <label htmlFor="ageRange">
          Age Range:
          <br />
          <br />
          <select
            id="ageRange"
            name="ageRange"
            type="number"
            value={fields.ageRange}
            onChange={handleFieldChange}
          >
            <option value="select age range">Select Age Range</option>
            <option value="0-3">0-3 Years</option>
            <option value="3-6">3-6 Years</option>
            <option value="6-9">6-9 Years</option>
            <option value="9-12">9-12 Years</option>
            <option value="12+">12+</option>
          </select>
        </label>
        <label htmlFor="description">
          Description:
          <br />
          <br />
          <input
            id="description"
            name="description"
            type="text"
            placeholder="Enter description here..."
            value={fields.description}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="borrow-period">
          Borrow Period:
          <br />
          <br />
          <select
            id="borrowPeriod"
            name="borrowPeriod"
            type="number"
            placeholder="select borrow period"
            value={fields.borrowPeriod}
            onChange={handleFieldChange}
          >
            <option value="select borrow period">Select Borrow Period</option>
            <option value="0-3">1 Month</option>
            <option value="3-6">3 Months</option>
            <option value="6-9">6 Months</option>
            <option value="9-12">12 Months</option>
            <option value="12+">12+ Months</option>
          </select>
        </label>
        <label htmlFor="postcode">
          Postcode:
          <br />
          <input
            type="text"
            name="postcode"
            id="postcode"
            placeholder="Enter postcode"
            value={fields.postcode}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="image">
          Image Upload:
          <br />
          <input
            type="text"
            name="image"
            id="image"
            placeholder="Enter demo image name"
            value={fields.image}
            onChange={handleFieldChange}
          />
        </label>
        <button className="button" type="submit" disabled={pristine}>
          Add
        </button>
      </form>
      <Alert
        message={alert.message}
        success={alert.isSuccess}
        onChange={handleFieldChange}
      />
    </div>
  );
}

export default AddToys;
