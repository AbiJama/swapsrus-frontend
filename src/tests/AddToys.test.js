import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import AddToys from "../components/AddToys";

describe("AddToys", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <AddToys />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
    const toyTitleInput = screen.getByLabelText(/Toy Title/i);

    expect(toyTitleInput).toBeInTheDocument();
  });
});
