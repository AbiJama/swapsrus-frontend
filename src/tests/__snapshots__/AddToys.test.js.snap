// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToys renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-toys"
  >
    <h2>
      Add Toys
    </h2>
    <form
      class="add-toy-form"
    >
      <label
        for="title"
      >
        Toy Title:
        <br />
        <br />
        <input
          id="title"
          name="title"
          placeholder="Enter title here..."
          value=""
        />
      </label>
      <label
        for="type"
      >
        Type:
        <br />
        <br />
        <select
          id="type"
          name="type"
        >
          <option
            value="Choose Category"
          >
            Choose Category
          </option>
          <option
            value="Books"
          >
            Books
          </option>
          <option
            value="Pre-school"
          >
            Pre-school
          </option>
          <option
            value="Indoor"
          >
            Indoor
          </option>
          <option
            value="Outdoor"
          >
            Outdoor
          </option>
        </select>
      </label>
      <label
        for="condition"
      >
        Condition:
        <br />
        <br />
        <select
          id="condition"
          name="condition"
        >
          <option
            value="Select Condition"
          >
            Select Condition
          </option>
          <option
            value="Brand New"
          >
            Brand New
          </option>
          <option
            value="Like New"
          >
            Like New
          </option>
          <option
            value="Good"
          >
            Good
          </option>
          <option
            value="Fair"
          >
            Fair
          </option>
          <option
            value="Defected"
          >
            Defected
          </option>
        </select>
      </label>
      <label
        for="ageRange"
      >
        Age Range:
        <br />
        <br />
        <select
          id="ageRange"
          name="ageRange"
          type="number"
        >
          <option
            value="select age range"
          >
            Select age range
          </option>
          <option
            value="0-3"
          >
            0-3
          </option>
          <option
            value="3-6"
          >
            3-6
          </option>
          <option
            value="6-9"
          >
            6-9
          </option>
          <option
            value="9-12"
          >
            9-12
          </option>
          <option
            value="12+"
          >
            12+
          </option>
        </select>
      </label>
      <label
        for="description"
      >
        Description:
        <br />
        <br />
        <input
          id="description"
          name="description"
          placeholder="Enter description here..."
          type="text"
          value=""
        />
      </label>
      <label
        for="borrow-period"
      >
        Borrow Period:
        <br />
        <br />
        <select
          id="borrowPeriod"
          name="borrowPeriod"
          placeholder="0"
          type="number"
        >
          <option
            value="select borrow period"
          >
            Select Borrow Period
          </option>
          <option
            value="0-3"
          >
            1 month
          </option>
          <option
            value="3-6"
          >
            3 months
          </option>
          <option
            value="6-9"
          >
            6 months
          </option>
          <option
            value="9-12"
          >
            12 months
          </option>
          <option
            value="12+"
          >
            12+months
          </option>
        </select>
      </label>
      <label
        for="postcode"
      >
        Postcode:
        <br />
        <input
          id="postcode"
          name="postcode"
          placeholder="Enter postcode"
          type="text"
          value=""
        />
      </label>
      <label
        for="image"
      >
        Image Upload:
        <br />
        <input
          id="image"
          name="image"
          placeholder="Enter demo image name"
          type="text"
          value=""
        />
      </label>
      <button
        class="button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
